xarray/core/variable.py

```diff
@@ -1382,107 +1381,6 @@ def roll(self, shifts=None, **shifts_kwargs):
-            What to do if dimensions that should be selected from are not present in the
-            Variable:
-            - "raise": raise an exception
-            - "warn": raise a warning, and ignore the missing dimensions
-            - "ignore": ignore the missing dimensions
-
-        Returns
-        -------
-        transposed : Variable
-            The returned object has transposed data and dimensions with the
-            same attributes as the original.
-
-        Notes
-        -----
-        This operation returns a view of this variable's data. It is
-        lazy for dask-backed Variables but not for numpy-backed Variables.
-
-        See Also
-        --------
-        numpy.transpose
-        """
-        if len(dims) == 0:
-            dims = self.dims[::-1]
-        else:
-            dims = tuple(infix_dims(dims, self.dims, missing_dims))
-
-        if len(dims) < 2 or dims == self.dims:
-            # no need to transpose if only one dimension
-            # or dims are in same order
-            return self.copy(deep=False)
-
-        axes = self.get_axis_num(dims)
-        data = as_indexable(self._data).transpose(axes)
-        return self._replace(dims=dims, data=data)
-
-    @property
-    def T(self) -> Self:
-        return self.transpose()
-
-    def set_dims(self, dims, shape=None):
-        """Return a new variable with given set of dimensions.
-        This method might be used to attach new dimension(s) to variable.
-
-        When possible, this operation does not copy this variable's data.
-
-        Parameters
-        ----------
-        dims : str or sequence of str or dict
-            Dimensions to include on the new variable. If a dict, values are
-            used to provide the sizes of new dimensions; otherwise, new
-            dimensions are inserted with length 1.
-
-        Returns
-        -------
-        Variable
-        """
-        if isinstance(dims, str):
-            dims = [dims]
-
-        if shape is None and utils.is_dict_like(dims):
-            shape = dims.values()
-
-        missing_dims = set(self.dims) - set(dims)
-        if missing_dims:
-            raise ValueError(
-                f"new dimensions {dims!r} must be a superset of "
-                f"existing dimensions {self.dims!r}"
-            )
-
-        self_dims = set(self.dims)
-        expanded_dims = tuple(d for d in dims if d not in self_dims) + self.dims
-
-        if self.dims == expanded_dims:
-            # don't use broadcast_to unless necessary so the result remains
-            # writeable if possible
-            expanded_data = self.data
-        elif shape is not None:
-            dims_map = dict(zip(dims, shape))
-            tmp_shape = tuple(dims_map[d] for d in expanded_dims)
-            expanded_data = duck_array_ops.broadcast_to(self.data, tmp_shape)
-        else:
-            expanded_data = self.data[(None,) * (len(expanded_dims) - self.ndim)]
-
-        expanded_var = Variable(
```

Comment:
@dcherian, I'm a bit confused. The existing code seems to allow set_dims on indexVariable and returns a Variable. Is the behavior of always returning Variable() in set_dims intentional?

- before 
```python
In [1]: import xarray as xr, numpy as np

In [2]: v = xr.Variable(['x'], [1, 2, 3])

In [3]: v.to_index_variable().set_dims(['x', 'y'])
Out[3]: 
<xarray.Variable (x: 3, y: 1)>
array([[1],
       [2],
       [3]])
```

- after 
```python
In [1]: import xarray as xr, numpy as np

In [2]: n = xr.namedarray.core.NamedArray(['x', 'y'], np.array([[99, 100], [1, 2]]))

In [3]: n.set_dims(['x', 'y', 'z'])
Out[3]: 
<xarray.NamedArray (x: 2, y: 2, z: 1)>
array([[[ 99],
        [100]],

       [[  1],
        [  2]]])

In [4]: v = xr.Variable(['x'], [1, 2, 3])

In [5]: v.set_dims(['x', 'y'])
Out[5]: 
<xarray.Variable (x: 3, y: 1)>
array([[1],
       [2],
       [3]])

In [6]: v.to_index_variable()
Out[6]: 
<xarray.IndexVariable 'x' (x: 3)>
array([1, 2, 3])

In [7]: v.to_index_variable().set_dims(['x', 'y'])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[7], line 1
----> 1 v.to_index_variable().set_dims(['x', 'y'])

File ~/devel/pydata/xarray/xarray/namedarray/core.py:958, in NamedArray.set_dims(self, dims, shape)
    955 else:
    956     expanded_data = self.data[(None,) * (len(expanded_dims) - self.ndim)]
--> 958 expanded_obj = self._replace(data=expanded_data, dims=expanded_dims)
    959 return expanded_obj.transpose(*dims)

File ~/devel/pydata/xarray/xarray/core/variable.py:964, in Variable._replace(self, dims, data, attrs, encoding)
    962 if encoding is _default:
    963     encoding = copy.copy(self._encoding)
--> 964 return type(self)(dims, data, attrs, encoding, fastpath=True)

File ~/devel/pydata/xarray/xarray/core/variable.py:2516, in IndexVariable.__init__(self, dims, data, attrs, encoding, fastpath)
   2514 super().__init__(dims, data, attrs, encoding, fastpath)
   2515 if self.ndim != 1:
-> 2516     raise ValueError(f"{type(self).__name__} objects must be 1-dimensional")
   2518 # Unlike in Variable, always eagerly load values into memory
   2519 if not isinstance(self._data, PandasIndexingAdapter):

ValueError: IndexVariable objects must be 1-dimensional
```

Reply:
i decided to keep this feature in place by space casing the return value in  `IndexVariable.set_dims()`